Problem for Graphs week: Power stations (#212)
UWB-ACM/CTCI: 2020-02-Spring/4_graphs/README.md

<!-- Don't remove -->
<a name="p2"/>

> ### 2. PROBLEM 2

Reviewer: Problem name

> ### 2. PROBLEM 2

Reviewer 2: @mduw move this to Problem 3; it seems complex enough. I have something simpler in mind and can do Problem 1

Author: fixed!

#### Scenario

> The electrical system of HappyLand consists of ``N`` stations from ``0`` to ``N-1`` including ``M`` power plants where electricity is produced and the rest ``N-M`` stations are transmission substations, ``M <= N``. After a hurricane, HappyLand’s authorities need to turn off some stations for maintenance. For safety reasons, an offline station has no electricity coming in/out of it. Consequently, substations rely only on offline stations are disconnected from the grid. Power plants can power themselves and adjacent stations. **Find out which stations are shutdown by disconnecting K stations.**

Reviewer 3: I feel like there's a grammar mistake in this sentence that makes it confusing:
`Consequently, substations rely only on offline stations are disconnected from the grid.`

Reviewer: I too find this confusing

Author: thanks guys. pls take a look at it again. let me know if it needs to be changed

Each station is represented by a vertex. Directed arrows indicates the electrical dependency between stations (receive from or transfer to its neighboring stations). Each substation is powered by at least one power plant.
<details>

> <summary>Click to see example input</summary>

Reviewer: This should not be hidden.

![Electrical system](./PowerStations/Figure12.png)

#### Example Input
```
N =6; M=2; K=1;
PowPlants[M] = {0,5};
Maint[N] = {0};
Matrix[N][N] = {{0,1,0,1,0,0},
              {0,0,1,0,1,0},
              {0,0,0,0,0,0},
              {0,0,0,0,1,0},
              {0,0,1,0,0,0},
              {0,0,0,0,1,0}};
```
#### Example Output
Offline stations:
```
0 1 3
```

> ##### Explain:

Reviewer: ```suggestion
##### Explanation:
```

<!-- Don't remove -->
<a name="s2"/>

> ### 2. SOLUTION 2

Reviewer: Problem name

> Instead of validating the status of current vertex by checking vertices(stations) it relies on, we could approach the problem top-down: we assume that all stations are shut down. Starting with known power plant thats remain active, we activate adjacent stations and substations of the current active station using DFS.Stations that are not activated by the end of the algorithm are the ones affected by shutting down K stations.

Reviewer: ```suggestion
Instead of validating the status of current vertex by checking the vertices(stations) it relies on, we could approach the problem top-down: we assume that all stations are shut down. Starting with known power plant that remains active, we activate adjacent stations and substations of the current active station using DFS. Stations that are not activated by the end of the algorithm are the ones affected by shutting down K stations.
```

<details>
<summary>Click to see the algorithm illustration</summary>

![Algorithm in steps](./PowerStations/Alg_illustration.png)
</details>

###### Pseudocode
```
void Solve() {
  -> mark all stations as Offline
  -> mark power plants that are in maintenance as Offline
  // start marking from active power plants i-th
  for (i : M) {
    if (PowPlants[i] is active)
      DFS(PowPlants[i]);
  }
}
void DFS(int Cur) {
  if (Cur is in maintenance) return; // stop checking

  // Otherwise mark Cur as active and visited
  Visited[Cur] = true;
  Active[Cur] = true;

  // check its adjacent
  for (i : N) {
    // if i is not visited and has path from Cur->i
    if (!Visited[i] && Matrix[Cur][i])
      DFS(i);
  }
}
```

> ###### Source code

Reviewer: Personally I think the code could use more comments

> ###### Source code

Reviewer: Source code indentation is off. If this code were to be copy/pasted and compiled, it likely wouldn't work. See your solution files to see what it should look like.

private static void DFS(Graph G, int Cur, boolean ActiveStas[], boolean Visited[]) {
  if (Maint[Cur]) return;
  Visited[Cur] = true;
  ActiveStas[Cur] = true;
  for (int i=0; i<G.N; i++) {
    if (i!=Cur && !Visited[i] && G.Matrix[Cur][i] == 1) {
      DFS(G, i, ActiveStas, Visited);
    }
  }
}
public static void Solve(Graph G) {
  // init
  Maint = new boolean[G.N];
  PowPlants = new boolean[G.N];
  for (int i=0; i<G.K; i++)
    Maint[G.Maint[i]] = true;
  for (int i=0; i<G.M; i++)
    PowPlants[G.PowPlants[i]] = true;

  // assuming all stations are off except active power plants
  boolean ActiveStas[] = new boolean[G.N];
  boolean Visited[] = new boolean[G.N];
  for (int i=0; i<G.N; i++) {
    ActiveStas[i] = false;
    if (PowPlants[i] && !Maint[i]) ActiveStas[i] = true;
      Visited[i] = false;
  }
  // using DFS to mark stations that connected to
  // known active stations
  for (int i=0; i<G.M; i++) {
    if (ActiveStas[G.PowPlants[i]])
      DFS(G, G.PowPlants[i], ActiveStas, Visited);
  }

  // print out offline stations at the end
  for (int i=0;i<G.N; i++) {
    if (ActiveStas[i]) continue;
    System.out.print(i +" ");
  }
  System.out.println();
}
```

</details>

> ###### Complexity:

Reviewer: Is this space or time complexity? It needs to be labeled. Add the other if you know it/can find it,

- Best and average case: when M is significantly small to N, it is ``O(M*N)``.

> - Worst case: when M is close to N or M==N, the algorithm has the sam complexity as that of dfs which is ``O(N^2)``.

Reviewer: ```suggestion
- Worst case: when M is close to N or M==N, the algorithm has the same complexity as that of DFS which is ``O(N^2)``.
```
